include/revng/ADT/GenericGraph.h

```diff
@@ -353,11 +370,557 @@ class BidirectionalNode
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToConstView>;
+  using PredecessorIterator = revng::mapped_iterator<
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToNeighbor>;
+  using ConstPredecessorIterator = revng::mapped_iterator<
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToConstNeighbor>;
+
+public:
+  llvm::iterator_range<SuccessorEdgeIterator> successor_edges() {
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return revng::map_range(Range, SuccessorFilters::ToView);
+  }
+  llvm::iterator_range<ConstSuccessorEdgeIterator> successor_edges() const {
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return revng::map_range(Range, SuccessorFilters::ToConstView);
+  }
+  llvm::iterator_range<SuccessorIterator> successors() {
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return revng::map_range(Range, SuccessorFilters::ToNeighbor);
+  }
+  llvm::iterator_range<ConstSuccessorIterator> successors() const {
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return revng::map_range(Range, SuccessorFilters::ToConstNeighbor);
+  }
+
+  llvm::iterator_range<PredecessorEdgeIterator> predecessor_edges() {
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return revng::map_range(Range, PredecessorFilters::ToView);
+  }
+  llvm::iterator_range<ConstPredecessorEdgeIterator> predecessor_edges() const {
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return revng::map_range(Range, PredecessorFilters::ToConstView);
+  }
+  llvm::iterator_range<PredecessorIterator> predecessors() {
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return revng::map_range(Range, PredecessorFilters::ToNeighbor);
+  }
+  llvm::iterator_range<ConstPredecessorIterator> predecessors() const {
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return revng::map_range(Range, PredecessorFilters::ToConstNeighbor);
+  }
+
+protected:
+  auto findSuccessorImpl(DerivedType const *S) {
+    auto Comparator = [S](auto &Edge) { return Edge.Neighbor == S; };
+    return std::find_if(Successors.begin(), Successors.end(), Comparator);
+  }
+  auto findSuccessorImpl(DerivedType const *S) const {
+    auto Comparator = [S](auto const &Edge) { return Edge.Neighbor == S; };
+    return std::find_if(Successors.begin(), Successors.end(), Comparator);
+  }
+  auto findPredecessorImpl(DerivedType const *P) {
+    auto Comparator = [P](auto &Edge) { return Edge.Neighbor == P; };
+    return std::find_if(Predecessors.begin(), Predecessors.end(), Comparator);
+  }
+  auto findPredecessorImpl(DerivedType const *P) const {
+    auto Comparator = [P](auto const &Edge) { return Edge.Neighbor == P; };
+    return std::find_if(Predecessors.begin(), Predecessors.end(), Comparator);
+  }
```

Comment:
Well, `const` and non-`const` need to be separate in any case, because that have different access requirements.

And `successor` cases are not really related to the `predecessor` ones as they use different containers.

That's how we come to 4 unrelated-ish methods. Maybe there's a better way, but I can't think of it.

Reply:
I'm thinking of a template method that takes as argument either `Successors` or `Predecessors` and is called by all the existing methods.

Very rough idea:

```c++
template<typename T>
auto impl(T &Neighbors) {
    auto Comparator = [P](auto &Edge) { return Edge.Neighbor == P; };
    return std::find_if(Neighbors.begin(), Neighbors.end(), Comparator); 
}

auto findPredecessorImpl(DerivedType const *P) const {
  return impl(Predecessors);
}

// ...
```

Reply:
Yes, of course, I don't mind something like this. Will try it out tomorrow. Sorry I didn't think of this myself.

Reply:
The problem is, this still kind of forces us to have two `Impl` function, a `const` one and a non-`const` one.

How much would you hate something along [these lines](https://stackoverflow.com/a/856839/8216655)?

Reply:
The main problem is that for me, 
```cpp
template<typename ContainerType>
auto findImpl(DerivedType const *What, ContainerType &&Where) {
  auto Comparator = [What](auto const &Edge) { return Edge.Neighbor == What; };
  return std::find_if(Successors.begin(), Successors.end(), Comparator);
}
template<typename ContainerType>
auto findImpl(DerivedType const *What, ContainerType &&Where) const {
  auto Comparator = [What](auto const &Edge) { return Edge.Neighbor == What; };
  return std::find_if(Successors.begin(), Successors.end(), Comparator);
}

auto findSuccessorImpl(DerivedType const *Successor) {
  return findImpl(Successor, Successors);
}
auto findSuccessorImpl(DerivedType const *Successor) const {
  return findImpl(Successor, Successors);
}
auto findPredecessorImpl(DerivedType const *Predecessor) {
  return findImpl(Predecessor, Predecessors);
}
auto findPredecessorImpl(DerivedType const *Predecessor) const {
  return findImpl(Predecessor, Predecessors);
}
```
is not any better than
```cpp
auto findSuccessorImpl(DerivedType const *S) {
  auto Comparator = [S](auto &Edge) { return Edge.Neighbor == S; };
  return std::find_if(Successors.begin(), Successors.end(), Comparator);
}
auto findSuccessorImpl(DerivedType const *S) const {
  auto Comparator = [S](auto const &Edge) { return Edge.Neighbor == S; };
  return std::find_if(Successors.begin(), Successors.end(), Comparator);
}
auto findPredecessorImpl(DerivedType const *P) {
  auto Comparator = [P](auto &Edge) { return Edge.Neighbor == P; };
  return std::find_if(Predecessors.begin(), Predecessors.end(), Comparator);
}
auto findPredecessorImpl(DerivedType const *P) const {
  auto Comparator = [P](auto const &Edge) { return Edge.Neighbor == P; };
  return std::find_if(Predecessors.begin(), Predecessors.end(), Comparator);
}
```

The second one at least looks more explicit.

Reply:
`findImpl` can be `static` I guess. That we you don't need two versions.

Reply:
Oh, wait, right, I'm stupid. Sorry and thanks :)

Reply:
What do you think now?